src/scheduler/node_info.c

```diff
@@ -1213,20 +1457,67 @@ dup_nodes(node_info **onodes, server_info *nsinfo,
 	}
 #endif /* localmod 049 */
 
-	for (i = 0; onodes[i] != NULL; i++) {
-		if ((nnodes[i] = dup_node_info(onodes[i], nsinfo, flags)) == NULL) {
-			free_nodes(nnodes);
-			return NULL;
+	tid = *(int *) pthread_getspecific(th_id_key);
+	if (tid != 0) { /* worker thread, don't use multithreading */
+		tdata = malloc(sizeof(th_data_dup_nd_info));
+		tdata->flags = flags;
+		tdata->nsinfo = nsinfo;
+		tdata->onodes = onodes;
+		tdata->nnodes = nnodes;
+		tdata->sidx = 0;
+		tdata->eidx = num_nodes - 1;
+		dup_node_info_chunk(tdata);
+		malloc_err = tdata->malloc_err;
+		free(tdata);
+	} else { /* We are multithreading */
+		j = 0;
+		chunk_size = num_nodes / num_threads;
+		chunk_size = (chunk_size > 1024)? chunk_size : 1024;
+		while (num_nodes_cpy > 0) {
+			num_tasks++;
+			tdata = malloc(sizeof(th_data_dup_nd_info));
+			tdata->flags = flags;
+			tdata->nsinfo = nsinfo;
+			tdata->onodes = onodes;
+			tdata->nnodes = nnodes;
+			tdata->sidx = j;
+			j += chunk_size;
+			tdata->eidx = j - 1;
+			task = malloc(sizeof(th_task_info));
+			task->task_type = TS_DUP_ND_INFO;
+			task->thread_data = (void *) tdata;
+
+			pthread_mutex_lock(&work_lock);
+			schd_enqueue(work_queue, (void *) task);
+			pthread_cond_signal(&work_cond);
+			pthread_mutex_unlock(&work_lock);
+
+			num_nodes_cpy -= chunk_size;
 		}
 
-#ifdef NAS /* localmod 049 */
-		if (allocNASrank) {
-			nnodes[i]->NASrank = onodes[i]->NASrank;
-			nsinfo->nodes_by_NASrank[onodes[i]->NASrank] = nnodes[i];
+		/* Get results from worker threads */
+		for (i = 0; i < num_tasks;) {
+			pthread_mutex_lock(&result_lock);
+			while (schd_is_empty(result_queue))
+				pthread_cond_wait(&result_cond, &result_lock);
+			pthread_mutex_unlock(&result_lock);
+			while (!schd_is_empty(result_queue)) {
+				task = (th_task_info *) schd_dequeue(result_queue);
+				tdata = (th_data_dup_nd_info *) task->thread_data;
+				malloc_err |= tdata->malloc_err;
+				free(tdata);
+				free(task);
+				i++;
+			}
```

Comment:
There isn't really much that the worker threads leave for the main thread to do in this case. Worker threads directly modify the new node info array for their respective indices and dup data into it. The only thing of use for the main thread is the error bit. It still has to loop over the result queue, wait until all worker threads post their results, and only then move on.

Reply:
Interesting.  I guess this just didn't match up with how I imagined how threads worked.  I thought you had some master process that spun up threads and then collected results and processed them.  I guess this works just as well.  The threads themselves know what portion of the problem they are working on, and place the results back in the right place.  One downside to this is if threads ever get confused on portion of the problem they are working on and overlap with another thread, it will be really hard to debug.  Debugging threads is going to be difficult in any case.  Just curious, what made you go with this model instead of the master/client model?

Reply:
I just went for the most efficient way. If we went with the master thread assembling the different dup'd fractions, it'd be worse w.r.t. performance and memory as the worker threads would send their results in their own dup arrays, and the master thread would iterate over their individual results and collate them. If the worker threads got confused, we'd see a mess in the final array regardless of which approach we go with. It's a very simple partitioning logic though, so they really shouldn't get confused.

Reply:
If you go with Arun's suggestion to make the thread pool code into a generic library, would that change your mind?  I don't know enough about threads to know one way or the other.

Reply:
Depending on the task, the worker threads might do very different things. So, even if we made it generic, depending on the task at hand, the master and worker threads will still have different levels of responsibilities, the master thread may do more post-processing for some tasks and less for others.

Reply:
Yeah we'd have to use callbacks to reassemble things as well.